lib/ex_doc/formatter/html.ex

```diff
@@ -79,6 +83,113 @@ defmodule ExDoc.Formatter.HTML do
     end
   end
 
+  defp group_by_nesting([], _), do: []
+
+  defp group_by_nesting(nodes, %{group_modules_by_nesting: []}), do: nodes
+
+  defp group_by_nesting(nodes, %{group_modules_by_nesting: prefixes}) do
+    # At this stage, nodes have already been sorted by group and id. We preserve the
+    # ordering of the processed nodes as we insert new ones, since the inserted context
+    # nodes have no `id` property. This enables the sidebar UI code to determine if a
+    # node is only for context (because it has no `id`), but prevents sorting nodes
+    # again using the `GroupMatcher` as it relies in part on the `id`.
+
+    %{group: g} = hd(nodes)
+
+    # * `context_group` tracks the current module group (as defined in `:groups_for_modules`)
+    #   we're processing: we want to treat each group independently with respect to displaying
+    #   module prefixes for nesting.
+    # * `nesting_context` is the module name of the module node we most recently
+    #   inserted to provide context about the nested modules. Note that the same module name
+    #   might have to be inserted multiple times as the context changes.
+    # * `processed_nodes` accumulates processed nodes (and newly inserted nesting context nodes)
+    #   across all groups.
+    acc = %{
+      context_group: g,
+      nesting_context: nil,
+      prefixes_to_nest: prefixes,
+      processed_nodes: []
+    }
+
+    Enum.reduce(nodes, acc, &process_unnested_node/2)
+    |> Map.get(:processed_nodes)
+    |> Enum.reverse()
+  end
+
+  defp process_unnested_node(%ModuleNode{} = module_node, acc) do
+    module_node
+    |> maybe_truncate_node_title(acc.prefixes_to_nest)
+    |> maybe_insert_nesting_context_node(acc)
```

Comment:
I wonder if we should rather insert those nodes in the javascript side, like we do for groups. It is relatively easier to do it there. From the Elixir side, we will emit JS nodes with two information: the nesting and the nested_title. In the JS side, we check if the current node has a nesting, if so and it is different than the previous nesting, then we output the nesting header.

Reply:
I'd have to check if we can ensure consistent results if we do it on the js side. In particular, the ["exceptions" section](https://davidsulc.github.io/timex_docs/Timex.Format.FormatError.html) gets nesting "for free" by doing the node insertion in the elixir code. I.e. it's completely transparent whether a node is a module, an exception, or whether it belongs to a group: the nesting will behave the same. I need to check whether there's centralized processing in the js code or not (haven't really dived into that yet).

Reply:
It is exactly the sidebar_items.handlebars that you have changed. Here is the code for the group: https://github.com/elixir-lang/ex_doc/pull/907/files#diff-d85f51765b385ca15bb7242d5936ac89R3

I propose something like very similar to that, in pseudo code it would be:

```
    {{#newNesting ../this node.nesting}}
      <li class="nesting">{{node.nesting}}</li>
    {{/newNesting}}
```

However, keep in mind that, if we start a new group and the first item of the group has a nesting, we need to emit the nesting too. So maybe you also need to pass the group as argument. Alternatively, when you check for newGroup and the group change, you just reset the parent "last nesting" value.